raft.go

```diff
@@ -1375,7 +1375,19 @@ func (r *Raft) appendEntries(rpc RPC, a *AppendEntriesRequest) {
 		rpc.Respond(resp, rpcErr)
 	}()
 
-	// Ignore an older term
+	// this use case would happen when a node was a voter and is added back to the cluster as non voter,
+	// it term could be higher then the cluster, but because it's a non voter that term need to be discarded
+	// in favor of the cluster current term to keep the cluster stable, as an election don't need to be started
+	// by a node which don't have voting rights.
+	currentTerm := r.getCurrentTerm()
+	hasVote := hasVote(r.getLatestConfiguration(), r.localID)
+	if a.Term < currentTerm && !hasVote {
+		r.setState(Follower)
+		r.setCurrentTerm(a.Term)
+		resp.Term = a.Term
+	}
```

Comment:
> if the non-voter had a record of the last term it saw a stable leader in

🤔 actually that might be the same thing as the term of it's last persisted log so we _do_ have that... but I'm not sure how we use it correctly. Above I was thinking we could reset the non-voters current Term to that but I think that's dangerous as noted above, perhaps we could use this value in the new check here though and only downgrade the term if a non-voter AND on a term higher than we've ever seen an append happen at before? That would safely fence out the old leader case I think.

Reply:
And I think it's not going to help either because in this case the non-voter isn't disrupting by trying to get votes with a higher term (it's a non-voter after all) but rather returning it during replication.

Reply:
So I think the problem can only occur if:

 1. The non-voter starts up with a `currentTerm` persisted that is higher than the actual stable leader term.
     * I assume this can only happen when a voter is restarted as a non-voter and happens to attempt an election just before it restarts causing it to increment its term. It has to persist the new term to it's state file, but not actually deliver the RPCs (since that would cause an election and a move to that term or higher in the rest of the cluster anyway)
  2. When it starts up, it's `currentTerm` is `stableLeaderTerm + 1` (assuming no leader change since it shutdown which seems to be required for this bug to manifest) but its last appended log term will be at most `stableLeaderTerm` (could be lower in theory).
  3. When the stable leader attempts an append entries at `stableLeaderTerm` it rejects it sending it's higher term in response.
  4. (before this patch) leader steps down and a new election happens needlessly.

Does that match your understanding Dhia? Specifically what I said in 1 the only possible way to get into this state or did I miss something?

If so then I _think_ a solution would be:
 * If we get an `appendEntries` for a term lower than `currentTerm` AND we are a non-voter AND the term from the apparent leader is greater than or equal to the term of the last log entry we have persisted. Then it's safe to assume we are in this state.

I reason that this is safe because we _might_ still revert to a term of an old leader and append it's entries, but we'll never delete entries from a later term in doing it which means that we effectively are in the same state as if this node were partitioned along with the old leader for a while before it stepped down, and that state will be corrected as soon as the new leader makes contact.

We can't flip flop - if the new leader made contact first and appended in the new term then the additional check above will effectively fence out the old leader from modifying our log and we will proceed normally.

What do you think @dhiaayachi ?

Reply:
Thank you for reasoning with me about this @banks! I was trying to go through those scenario and I agree it's not easy without a formal proof. I ended up reasoning based on the following and using facts that should not vary based on timing or external events:

- The node in question is a non-voter 
- Any election triggered by a non voter is a disruption because that node cannot participate in it
- Term kept by non voters is always coming from current leader as it can't participate in an election so it can't increase its term by its own.

Based on those assumptions, If the term of a non voter node is higher than the cluster current term this  bring us to 2 possible scenarios.

1. That higher term come from another node that stepped as a leader but partitioned from the majority of nodes (the scenario you just described)
2. That higher term is a residual from the node being a voter formerly (described in #524)

I agree, first scenario is not ideal with the fix as the nonvoter will be in an unstable state until the cluster is restored but if we run the same scenario without the fix, the non voter node will be in most cases following the partitioned node as it will very likely have a higher term, which is not practical in real life situation.

Reply:
I can think of a scenario that's a bit different to above and breaks my assumption _slightly_.

If the server while it had been a voter had been partitioned from the other servers it might have sat there for arbitrarily long attempting elections and so may have a persisted term that is arbitrarily higher than the stable leader. It didn't disrupt the leader due to not being able to send it messages.

Some time after it restarts as a non-voter with this state, the partition heals and the bug manifests when the current leader attempts to replicate. Note that the rest of the cluster might have been through any number of leader elections in the time before the partition healed (either before or after the node restarted) but as long as the new leader's term is lower than the term the partitioned voter got itself to the bug can occur.

I don't think that changes the correctness of the solution above, but I think it's worth understanding the most general way that this state is possible!

Reply:
> Some time after it restarts as a non-voter with this state, the partition heals and the bug manifests when the current leader attempts to replicate. Note that the rest of the cluster might have been through any number of leader elections in the time before the partition healed (either before or after the node restarted) but as long as the new leader's term is lower than the term the partitioned voter got itself to the bug can occur.

That is exactly the bug described in #524 and fall under 2 in my last comment.